### DISCUSSION TITLE
ESP32 SD Card Error (OSError: 16)

### DISCUSSION THREAD

**Discussion by skysurf76** [micropython/micropython#9935] [General]
I followed the instructions for using an SD Card with ESP32 in the MicroPython documentation at...  https://docs.micropython.org/en/latest/esp32/quickref.html#sd-card

I got the following results...

>>> import machine, os
>>> sd = machine.SDCard(slot=2)
>>> os.mount(sd, '/sd')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: 16

After some searching I found the following...
https://github.com/micropython/micropython/issues/8290

Is there a fix to this, or is it an unresolved issue?

**Comment by karfas** [comment]
I don't think you did read #8290 to the end. The original issue there is solved (see https://github.com/micropython/micropython/issues/8290#issuecomment-1054208392).

**Reply by skysurf76** [reply]
Yep, sorry about that. I did read that post but I was skimming and it didn't hit me that the solution was to put freq as an argument in machine.sdcard().

Thanks for pointing out my error.

**Comment by MATTYGILO** [comment]
Another possibility to the error is if you have a folder called SD on the flash of your device

**Comment by sosi-deadeye** [comment]
> Do your job and FIX it.

Do you seriously think that's going to speed anything up?

The last comment was Aug 18, 2023.
Is there still an issue?

**Comment by jlsilicon** [comment]


**Reply by Josverl** [reply]
Thanks to @Tangerino that has shared their hard work you could find out in detail what the problem is in your code, device or wiring.
https://github.com/Tangerino/micropython-p4-test-suite

you may have overlooked part of the internet in your search.
but what do I know, i'v just been coding for 50 years and still earning day by day

**Reply by jlsilicon** [reply]
ok, sorry

**Comment by robert-hh** [comment]
I just tried a ESP32 board and it works with SDCard. Please explain;

- what did you try
- which board do you use?
- the probably very short test code that you tried.

**Comment by jlsilicon** [comment]
Sorry,
I thought this Git / issues - was for the Esp32-P4.

There seems to be an issue that the Esp32-P4 is generating Error 16 already in use) - for the recent P4 chips across the internet using SD Cards.

Ok, I see it.
Google turns this up under the list of links under :

[machine.SDCard() fails with OSError: 16 on ESP32-P4 ...](https://github.com/orgs/community/discussions/188887)
GitHub
https://github.com › orgs › community › discussions
Mar 7, 2026 — ESP32-P4 machine.SDCard() returns OSError: 16 (works on ESP32 DevKit v2) Description I am trying to use an SD card with machine.SDCard() on an ESP32-P4 ...
[3 answers](https://github.com/orgs/community/discussions/188887#discussioncomment-16069492)
 Top answer: 
Since you’ve confirmed the hardware is strictly wired for SDMMC and the standard software resets aren't clearing the OSError: 16, it appears we’ve ...
[ESP32 SDCard OSError: 16 #10581 - micropython - GitHub](https://github.com/orgs/micropython/discussions/10581)
Feb 1, 2023
[ESP32 SD Card Error (OSError: 16) #9935 - GitHub](https://github.com/orgs/micropython/discussions/9935)
Jun 13, 2023

- my confusion.

**Reply by Josverl** [reply]
@jlsilicon 
I do not think it is fair to try to put the blame on a search engine, They did not write your posts.

You claim to have 30 years of experience, to be honest  based on you communication style I assume that to be much less. 
Still everyone is entitled to have a bad day sometimes - But I do ask you be more considerate next time you ask for help.

**Reply by projectgus** [reply]
This discussion is about the original ESP32 chip and was started in 2022 (and the solution was provided by the submitter in one of the first comments). The ESP32-P4 chip was released in 2025. The error message is the same, but there is otherwise no connection.

There is an issue with accessing the SDCard on ESP32-P4 that's fixed in the nightly MicroPython builds, or can be worked around in the MicroPython v1.28 release. I've posted a reply on the other discussion (the one about ESP32-P4): https://github.com/orgs/community/discussions/188887#discussioncomment-17503642

As @Josverl says, it's understandable to be frustrated when things don't work but please be polite and respectful when interacting with us.

**Comment by projectgus** [comment]
Closing this discussion as the original question was answered in https://github.com/orgs/micropython/discussions/9935#discussioncomment-4129155
